File: hw7/Driver.java
//==========================================================================
// cs12wi20cu                         Homework 7              Tianling Luo
//--------------------------------------------------------------------------
// File: Driver.java
//
// Description: In this assignment, we will write the class implementation
// for UCSDStudent which includes all UCSDStudent methods that will be
// called by the Tree code.
//==========================================================================

import java.io.*;  
      
//=========================================================================
// class UCSDStudent
//
// Description: The UCSDStudent class is defined and it contains the member 
// functions toString, UCSDStudent, getName, equals and isGreaterThan to
// be called by the Tree.java code.
//
// Data Fields:
//     name (String)  - the name of UCSD student
//     studentnum (long)  - the student number of UCSD student
//
// Public functions:
//     UCSDStudent - store the name and student number in name and
//     		     Studentnum with input String and long
//     getName - get the name of the student
//     equals - check if a student name is equal to another one
//     isGreaterThan - to compare the name with base
//     toString - get the name and student number in string
//==========================================================================  
class UCSDStudent extends Base {  
      
        private String name;  // name of student
    private long studentnum;  // student number of student
      
        public UCSDStudent (String nm, long val) {  
            // assign the name
            name = new String (nm); 
            // assign the studentnum
            studentnum = val;  
        }  
  
        public String getName () {  
            // return the name
            return name;  
        }  

        public boolean equals (Object object) {  
            if (this == object)  
                return true;  
  
            if (!(object instanceof UCSDStudent))  
                return false;  
          
            UCSDStudent otherVar = (UCSDStudent) object;  
            // check if the names are the same
            return name.equals (otherVar.getName ());  
        }
        
        public boolean isGreaterThan (Base base) {  
            // return true if it is greater
            return (name.compareTo (base.getName ()) > 0) ? true : false;  
        }    
      
        public String toString () {
            // return the string with name and student name
            return "name:  " + name + "  studentnum:  " + studentnum;  
        }  
}  
      
public class Driver {  
        private static final short NULL = 0;  
  
        public static void main (String [] args) {  
              
            // initialize debug states
            Tree.debugOff();  
    
            // check command line options
            for (int index = 0; index < args.length; ++index) {  
                if (args[index].equals("-x"))  
                    Tree.debugOn();  
            }  
      
      
            // The real start of the code
            SymTab<UCSDStudent> symtab = 
            new SymTab<UCSDStudent>("UCSDStudentTree");  
            String buffer = null;
            char command;  
            long number = 0;  
      
            System.out.println ("Initial Symbol Table:\n" + symtab);  
      
            while (true) {
                command = NULL; // reset command each time in loop  
                System.out.print ("Please enter a command:  " +   
                            "((a)llocate, (i)nsert, (l)ookup, " +  
                            "(r)emove, (w)rite):  ");  
      
                try {  
                    command = MyLib.getchar ();  
                    MyLib.clrbuf (command); // get rid of return  
      
                    switch (command) {  
                    case 'a':  
                            System.out.print  
                            ("Please enter name of new Tree to " +  
                            "allocate:  ");  
                                   
                            buffer = MyLib.getline ();// formatted input  
                            symtab = new SymTab<UCSDStudent>(buffer);  
                            break;  
      
                    case 'i':  
                            System.out.print  
                            ("Please enter UCSD student name to insert:  ");  
      
                            buffer = MyLib.getline ();// formatted input  
      
                            System.out.print  
                                    ("Please enter UCSD student number:  ");  
      
                            number = MyLib.decin ();  
                            MyLib.clrbuf (command); // get rid of return  
     
                            // create student and place in symbol table  
                            symtab.insert(new UCSDStudent (buffer, number));  
                            break;  
   
                    case 'l': {   
                            UCSDStudent found;      // whether found or not  
     
                            System.out.print  
                            ("Please enter UCSD student name to lookup:  ");  
                            buffer = MyLib.getline ();// formatted input  
      
                            UCSDStudent stu = new UCSDStudent (buffer, 0);  
                            found = symtab.lookup (stu);  
                                      
                            if (found != null) {  
                                    System.out.println("Student found!");  
                                    System.out.println(found);  
                            }  
                            else  
                                    System.out.println ("student " + buffer  
                                            + " not there!");  
                            }  
                            break;  
                              
                    case 'r': {   
                            UCSDStudent removed; // data to be removed  
     
                            System.out.print  
                            ("Please enter UCSD student name to remove:  ");  
  
                            buffer = MyLib.getline ();  
   
                            UCSDStudent stu = new UCSDStudent (buffer, 0);  
      
                            removed = symtab.remove(stu);  
      
                            if (removed != null) {  
                                    System.out.println("Student removed!");   
                                    System.out.println(removed);  
                            }  
                            else  
                                    System.out.println ("student " + buffer  
                                            + " not there!");  
                            }  
                            break;  
      
                    case 'w':  
                            System.out.println("The Symbol Table " +  
                            "contains:\n" + symtab);  
                    }  
                    }  
                    catch (EOFException eof) {  
                            break;  
                    }  
            }  
            System.out.println("\nFinal Symbol Table:\n" + symtab);  
    }  
}  

File: hw7/Driver.c
//==========================================================================
// cs12wi20cu                         Homework 7              Tianling Luo
//--------------------------------------------------------------------------
// File: Driver.c
//
// Description: In this assignment, we will write the class implementation
// for UCSDStudent which includes all UCSDStudent methods that will be
// called by the Tree code.
//==========================================================================

#include <cstdlib>  
#include <iostream>  
#include <string.h>  
#include "Base.h"  
#include "SymTab.h"  
using namespace std;

#ifdef NULL  
#undef NULL  
#define NULL 0  
#endif  

//=========================================================================
// class UCSDStudent
//
// Description: The UCSDStudent class is defined and it contains the member
// functions that will be called by the Tree.c code.
//
// Data Fields:
//     name (char*)  - the name of UCSD student
//     studentnum (long)  - the student number of UCSD student
//
// Public functions:
//     UCSDStudent - store the name and student number in name and
//     		     Studentnum with input char* and long
//     ~UCSDStudent - to delete name
//     operator const char* - get the name of the student
//     operator long == - is equal function in c++
//     operator long > - isGreaterThan function in c++
//     Write - get the name and student number in string
//==========================================================================  
class UCSDStudent : public Base {
    char* name;
    long studentnum;
public:

    // constructor 
    UCSDStudent(char* nm, long sn = 0) :
        name(strdup(nm)), studentnum(sn) {}

    // destructor
    ~UCSDStudent(void) {
        // freeing
        free(name);
    }

    // to get the name of student
    operator const char* (void) const {
        return name;
    }

    // to check if equal
    long operator == (const Base& bbb) const {
        return !strcmp(name, bbb);
    }

    // to check if larger
    long operator > (const Base& bbb) const {
        return (strcmp(name, bbb) > 0) ? 1 : 0;
    }

    // write element in it as string
    ostream& Write(ostream& stream) const {
        return stream << "name:  " << name
            << "  studentnum:  " << studentnum;
    }
};

int main(int argc, char* const* argv) {
    char buffer[80];
    char command;
    long number;

    Tree::Set_Debug(0);

    if (argc != 1 && strcmp("-x", argv[1]) == 0) {
        Tree::Set_Debug(1);
    }

    SymTab* ST;
    ST = new SymTab("UCSDStudentTree");
    ST->Write(cout << "Initial Symbol Table:\n");

    while (cin) {
        command = NULL;     // reset command each time in loop  
        cout << "Please enter a command:  ((a)llocate, (i)nsert, " <<
            "(l)ookup, (r)emove, (w)rite):  ";
        cin >> command;

        switch (command) {

        case 'a':
            cout << "Please enter name of new Tree to allocate:  ";
            cin >> buffer;

            delete ST;
            ST = new SymTab(buffer);

            break;

        case 'i':
            cout << "Please enter UCSD student name to insert:  ";
            cin >> buffer;    // formatted input  

            cout << "Please enter UCSD student number:  ";
            cin >> number;

            // create student and place in symbol table  
            ST->Insert(new UCSDStudent(buffer, number));
            break;

        case 'l': {
            const Base* found; // whether found or not  

            cout << "Please enter UCSD student name to lookup:  ";
            cin >> buffer;    // formatted input  

            UCSDStudent stu(buffer, 0);
            found = ST->Lookup(&stu);

            if (found)
                found->Write(cout << "Student found!\n") << "\n";
            else
                cout << "student " << buffer << " not there!\n";
        }
                break;

        case 'r': {
            Base* removed; // data to be removed  

            cout << "Please enter UCSD student name to remove:  ";
            cin >> buffer;    // formatted input  

            UCSDStudent stu(buffer, 0);
            removed = ST->Remove(&stu);

            if (removed)
                removed->Write(cout << "Student removed!\n") << "\n";
            else
                cout << "student " << buffer << " not there!\n";
        }
                break;


        case 'w':
            ST->Write(cout << "The Symbol Table contains:\n");
        }
    }

    ST->Write(cout << "\nFinal Symbol Table:\n");
    delete ST;
}
